strider/trapi_throttle/throttle.py

```diff
@@ -0,0 +1,417 @@
+    @wraps(fcn)
+    async def wrapper(*args, **kwargs):
+        try:
+            return await fcn(*args, **kwargs)
+        except Exception as err:
+            traceback.print_exc()
+            raise
+    return wrapper
+
+
+async def anull(arg, *args, **kwargs):
+    """Do nothing, asynchronously."""
+    return arg
+
+
+class ThrottledServer():
+    """Throttled server."""
+
+    def __init__(
+        self,
+        id: str,
+        url: str,
+        request_qty: int,
+        request_duration: float,
+        *args, 
+        max_batch_size: Optional[int] = None,
+        timeout: float = 60.0,
+        preproc: Callable = anull,
+        postproc: Callable = anull,
+        logger: logging.Logger = None,
+        **kwargs,
+    ):
+        """Initialize."""
+        self.id = id
+        self.worker: Optional[Task] = None
+        self.request_queue = asyncio.PriorityQueue()
+        self.counter = itertools.count()
+        self.url = url
+        self.request_qty = request_qty
+        self.request_duration = datetime.timedelta(seconds=request_duration)
+        self.timeout = timeout
+        self.max_batch_size = max_batch_size
+        self.preproc = preproc
+        self.postproc = postproc
+        if logger is None:
+            logger = logging.getLogger(__name__)
+        self.logger = logger
+
+    @log_errors
+    async def process_batch(
+            self,
+    ):
+        """Set up a subscriber to process batching"""
+        # Initialize the TAT
+        #
+        # TAT = Theoretical Arrival Time
+        # When the next request should be sent
+        # to adhere to the rate limit.
+        #
+        # This is an implementation of the GCRA algorithm
+        # More information can be found here:
+        # https://dev.to/astagi/rate-limiting-using-python-and-redis-58gk
+        if self.request_qty > 0:
+            interval = self.request_duration / self.request_qty
+            tat = datetime.datetime.utcnow() + interval
+
+        while True:
+            # Get everything in the stream or wait for something to show up
+            priority, (request_id, payload, response_queue) = await self.request_queue.get()
+            priorities = {
+                request_id: priority
+            }
+            request_value_mapping = {
+                request_id: payload
+            }
+            response_queues = {
+                request_id: response_queue
+            }
+            while True:
+                if self.max_batch_size is not None and len(request_value_mapping) == self.max_batch_size:
+                    break
+                try:
+                    priority, (request_id, payload, response_queue) = self.request_queue.get_nowait()
+                except QueueEmpty:
+                    break
+                priorities[request_id] = priority
+                request_value_mapping[request_id] = payload
+                response_queues[request_id] = response_queue
+
+            LOGGER.debug(
```

Comment:
```suggestion
            self.logger.debug(
```
I think we should try to stick to local loggers and reduce use of `LOGGER`

Reply:
7ea00f7b3b4ae5f233179d37b44458401f603945